runtime/commonMain/src/kotlinx/serialization/ContextAware.kt

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2017-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization
+
+import kotlinx.serialization.modules.*
+import kotlin.reflect.*
+
+/**
+ * Descriptor which allows retrieving additional runtime information, if proper [SerialModule] is provided.
+ *
+ * Usually, descriptors are generated by compiler plugin and know all information statically.
+ * However, that is not the case for [UnionKind.CONTEXTUAL] and [PolymorphicKind.OPEN] kinds of descriptors,
+ * where actual serializer used for property can be determined only at runtime.
+ * This descriptor is used by serializers which represent contextual serialization and open polymorphism (namely,
+ * [ContextSerializer] and [PolymorphicSerializer]), so they can capture statically known [kClass] in this descriptor.
+ * For schema introspection purposes, this [kClass] can be used in [SerialModule] to retrieve actual descriptor in runtime.
+ *
+ * This descriptor must always has either [UnionKind.CONTEXTUAL] or [PolymorphicKind.OPEN] kind,
+ * and vice versa: if some [SerialDescriptor.kind] is either CONTEXTUAL or OPEN, that descriptor
+ * could be safely casted to [ContextAwareDescriptor].
```

Comment:
Is downcast really allowed?
E.g. what about `descriptor.nullable`?

Reply:
It is unclear what to do when kind is CONTEXTUAL, but descriptor is not `ContextDescriptor`. I'd say it is possible to put some efforts in this and make `.nullable` aware of this particular subclass. I'm not 100% sure we should pursue this; we can discuss this in DM if you'd like